### PR TITLE
merge into development

### PR DESCRIPTION
**Pull request by rolfvdhulst**

### CODE REVIEW

File: include/connection.hpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            static_assert(type_traits::is_serializable_v<T>, "T is not serializable to string in Connection::read_next()");
+            zmqpp::message msg;
+            std::string data{};
+            if (!socket.receive(msg, true)) {
+                return stx::Err(std::move(data));
+            }
+            msg >> data;
+            T object;
+            auto succ = object.ParseFromString(data);
+            if (succ) {
```
Comment (Yuhanun) [inline include/connection.hpp:33]: Could inline this maybe

File: include/interface.hpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            server.start();
+        }
+
+        void handle_incoming(std::shared_ptr<ix::ConnectionState>, ix::WebSocket&, const ix::WebSocketMessagePtr& msg) {
+            proto::UiSettings data;
+            if (!data.ParseFromString(msg->str)) {
+                std::cerr << "Something else than proto::UiSettings has been received by central from ui" << std::endl;
+                return;
+            }
+            functor(data);
```
Comment (Yuhanun) [inline include/interface.hpp:51]: Filthy hacks.

File: src/server.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    void Server::handle_ai_state(proto::ModuleState ok) {
+      //TODO: send state to interface and modules
+//      auto handshakes = module_handshakes.acquire();
+//      for (auto const& each : *handshakes) {
+//        *ok.add_handshakes() = each;
+//      }
+//        // send it to the interface
+//        roboteam_interface.acquire()->write(ok);
+//        // forward this state to all modules.
+//        modules.broadcast(ok);
```
Comment (Yuhanun) [inline src/server.cpp:44]: ?

Comment (Yuhanun) [inline src/server.cpp]: Why is this commented out?

File: src/server.cpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        //TODO: if settingschanged or every 0.5 seconds, send settings to AI
+    }
+
+    void Server::handle_modules() {
+        modules.run(std::ref(this->_run));
+    }
+
+    void Server::run() {
+        std::cout << "Constructing AI thread." << std::endl;
+        new (&ai_thread) Mutex{ std::thread([this]() {
```
Comment (Yuhanun) [inline src/server.cpp:59]: I'm still unsure about these placement new's, even though there's no const members, if we add one it's going to be hnggggg std::launder time

**Review by Yuhanun — APPROVED** [review]
LGTM

File: src/server.cpp (hunk 3 above)
Comment (rolfvdhulst) [reply src/server.cpp]: Ah, I did that when trying to find deadlocks. I'll turn it back on.

File: src/server.cpp (hunk 4 above)
Comment (rolfvdhulst) [reply src/server.cpp:59]: This is over my head TBH, I don't know every intricate detail of c++.

File: src/server.cpp (hunk 3 above)
Comment (rolfvdhulst) [reply src/server.cpp]: Oh, no, it turns out there's a reason I disabled it. The modules.broadcast() line blocks/locks for some reason, I don't know exactly why.

Comment (Yuhanun) [reply src/server.cpp]: I think it blocks when there's no connections.

Comment (rolfvdhulst) [reply src/server.cpp]: Hmm ok. We should just use non-blocking connections then. We can't rely on the existence of connections in general.

**Pull request merged**